spring-kafka/src/main/java/org/springframework/kafka/core/reactive/ReactiveKafkaConsumerTemplate.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.core.reactive;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.consumer.OffsetAndTimestamp;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.kafka.receiver.KafkaReceiver;
+import reactor.kafka.receiver.ReceiverOptions;
+import reactor.kafka.receiver.ReceiverRecord;
+import reactor.kafka.sender.TransactionManager;
+
+/**
+ * Reactive kafka consumer operations implementation.
+ *
+ * @param <K> the key type.
+ * @param <V> the value type.
+ *
+ * @author Mark Norkin
+ */
+public class ReactiveKafkaConsumerTemplate<K, V> {
+	private final KafkaReceiver<K, V> kafkaReceiver;
+
+	public ReactiveKafkaConsumerTemplate(ReceiverOptions<K, V> receiverOptions) {
+		this.kafkaReceiver = KafkaReceiver.create(receiverOptions);
+	}
+
+	public Flux<ReceiverRecord<K, V>> receive() {
+		return this.kafkaReceiver.receive();
+	}
+
+	public Flux<ConsumerRecord<K, V>> receiveAutoAck() {
+		return this.kafkaReceiver.receiveAutoAck().flatMap(Function.identity());
+	}
+
+	public Flux<ConsumerRecord<K, V>> receiveAtmostOnce() {
+		return this.kafkaReceiver.receiveAtmostOnce();
+	}
+
+	public Flux<Flux<ConsumerRecord<K, V>>> receiveExactlyOnce(TransactionManager transactionManager) {
```

Comment:
I don't remember if we discussed this, but why `Flux` of `Flux`es?
Can't we flatten them? 
Or if that is some Reactor Kafka specific type for records per partition, then let's return that one maybe!

Just explain me one more time "Why?"

Thanks

Reply:
Well, looking at the documentation of KafkaReceiver, particularly:

> Returns a {@link Flux} of consumer record batches that may be used for exactly once delivery semantics. 
> ...
> A new transaction is started for each inner Flux and it is the responsibility of the consuming application to commit or abort the transaction
> ...
> The next batch of consumer records will be delivered only after the previous flux terminates.

My understanding of it is that we can't flatten them.
It results in having one flattened Flux which will never terminate and so is not clear where transaction demarcation happens/should happen.

Reply:
Done